README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

## Installation

* Add this to your package's pubspec.yaml file:

```
dependencies: 
    json_to_widget
```

* You can install packages from the command line: with Flutter:

```
$ flutter packages get
```

### JsonWidget

```
JsonWidget(
    decorations: decorations,
    form: form,
    onChanged: (dynamic response) {
        this.response = response;
    },
    actionSave: (data) {
        print(data);
    },
    buttonSave: new Container(
        height: 40.0,
        color: Colors.blueAccent,
        child: Center(
            child: Text("Login",style:TextStyle(color: Colors.white,fontWeight: FontWeight.bold)),
        ),
    ),
),
```

### Attribute
* form (Type String) Your form in String
* onChanged (Type Function)(1 parameter) call the function every time a change in the form is made
* padding (Type Double)
* formMap (Type Map) Your form in Map 
* errorMessages(Type Map) change string for error of required
* validations(Type Map) to add validation (TextInput,Input, Password, Email or TextArea)
* decorations(Type Map) to add decoration (TextInput,Input, Password, Email or TextArea)
* buttonSave(Type Widget) (not RaisedButton problem in onClick)
* actionSave(Type Function) the function is called when you click on the widget buttonSave

### Form
* Create Form String
```
String formString = json.encode({
    'title': 'form example',
    'description':'',
    'autoValidated': true, //default false
    'fields': [
        ...
    ]
});
```

* Create Form Map
```
Map formMap = {
    'title': 'form example',
    'description':'',
    'autoValidated': true, //default false
    'fields': [
        ...
    ]
};
```

### Fields
* All fields has attribute labelHidden(default false)
* Important add key for all field for validation required


### TextInput or Input

### Types?
* Input
* Password
* Email (has default validation)
* TextArea
* TextInput

```
// Example for json string
// to start with a default value you can add the value attribute
  String formString = json.encode({
    'fields': [
        {
             'key': 'inputKey',
             'type': 'Input',
             'label': 'Hi Group',
             'placeholder': "Hi Group flutter",
             'required': true
        },
        {
             'key': 'inputKey',
             'type': 'Input',
             'label': 'Initial Value',
             'value': 'Hello'
             'required': true
        },
    ]
 });
// Example for json Map
// in Map has Attributes validation and decoration

// important to receive 2 parameters in function of Validation
String validationExample(field, value) {
   if (value.isEmpty) {
     return 'Please enter some text';
   }
   return null;
}

Map formMap = {
    'fields': [
        {
                'key': 'inputKey',
                'type': 'Input',
                'label': 'Hi Group',
                'placeholder': "Hi Group flutter",
                'validator': 'digitsOnly',
                'required': true,
                'decoration': InputDecoration(
                          prefixIcon: Icon(Icons.account_box),
                          border: OutlineInputBorder(),
                ),
                'validation':validationExample
        },
    ]
 };
```

### How can I place validations for my form String?
* In JsonWidget has attributes (validations, decorations)
* Important that each field has to has its unique key

```
Map decorations = {
    'inputKey': InputDecoration(
      labelText: "Enter your email",
      prefixIcon: Icon(Icons.account_box),
      border: OutlineInputBorder(),
    ),
  };
  
Map validations = {
    'inputKey': validationExample,
}

dynamic response;
new JsonWidget(
    decorations: decorations,
    validations: validations,
    form: formString,
    onChanged: (dynamic response) {
        this.response = response;
    },
    actionSave: (data) {
        print(data);
    },
    buttonSave: new Container(
        height: 40.0,
        color: Colors.blueAccent,
        child: Center(
            child: Text("Login",style:TextStyle(color: Colors.white,fontWeight: FontWeight.bold)),
        ,
    ),
)
```

### Radio
```
String formString = json.encode({
    'fields': [
         {
                'key': 'radiobutton1',
                'type': 'RadioButton',
                'label': 'Radio Button tests',
                'value': 2,
                'items': [
                  {
                    'label': "product 1",
                    'value': 1,
                  },
                  {
                    'label': "product 2",
                    'value': 2,
                  },
                  {
                    'label': "product 3",
                    'value': 3,
                  }
                ]
         },
    ],
 });
```

### Switch
```
String formString = json.encode({
    'fields': [
         {
                 'key': 'switch1',
                 'type': 'Switch',
                 'label': 'Switch test',
                 'value': false,
         },
    ],
 });
```

### Checkbox
```
String formString = json.encode({
    'fields': [
        {
                'key': 'checkbox1',
                'type': 'Checkbox',
                'label': 'Checkbox test',
                'items': [
                  {
                    'label': "product 1",
                    'value': true,
                  },
                  {
                    'label': "product 2",
                    'value': false,
                  },
                  {
                    'label': "product 3",
                    'value': false,
                  }
                ]
        }
    ],
 });
```

### Select (New Field)
```
String formString = json.encode({
    'fields': [
         {
                 'key': 'select1',
                 'type': 'Select',
                 'label': 'Select test',
                 'value':'product 1',
                 'items': [
                   {
                     'label': "product 1",
                     'value': "product 1",
                   },
                   {
                     'label': "product 2",
                     'value': "product 2",
                   },
                   {
                     'label': "product 3",
                     'value': "product 3",
                   }
                 ]
         }
    ],
 });
```

### when text is added to the TextField, add field called response
```
// initial form_send_email
[{"type":"Input","label":"Subject","placeholder":"Subject"},{"type":"TextArea","label":"Message","placeholder":"Content"}]

// add text (hi) in TextField Message, update dynamic response; and add field called response
[{type: Input, label: Subject, placeholder: Subject, value: hello}, {type: TextArea, label: Message, placeholder: Content, value: hi }]
```

### Json to Widget
* Import it Now in your Dart code, you can use
```
import 'package:json_to_widget/json_to_widget.dart';
```

## Usage
* TextField
```
String form = json.encode([
    {
      'type': 'Input',
      'title': 'Hi Group',
      'placeholder': "Hi Group flutter"
    },
    {
      'type': 'Password',
      'title': 'Password',
    },
    {
      'type': 'Email', 
      'title': 'Email test',
      'placeholder': "hola a todos"
    },
    {
      'type': 'TareaText',
      'title': 'TareaText test',
      'placeholder': "hola a todos"
    },
  ]);
```

* Radio
```
String form = json.encode([
    {
      'type': 'RadioButton',
      'title': 'Radio Button tests',
      'value': 2,
      'list': [
        {
          'title': "product 1",
          'value': 1,
        },
        {
          'title': "product 2",
          'value': 2,
        },
        {
          'title': "product 3",
          'value': 3,
        }
      ]
    },
  ]);
```

* Switch
```
String form = json.encode([
    {
      'type': 'Switch',
      'title': 'Switch test',
      'switchValue': false,
    },
  ]);
```

* Checkbox
```
String form = json.encode([
    {
      'type': 'Checkbox',
      'title': 'Checkbox test 2',
      'list': [
        {
          'title': "product 1",
          'value': true,
        },
        {
          'title': "product 2",
          'value': true,
        },
        {
          'title': "product 3",
          'value': false,
        }
      ]
    },
  ]);
```

## Example
```
String form_send_email = json.encode([
    {'type': 'Input', 'title': 'Subject', 'placeholder': "Subject"},
    {'type': 'TareaText', 'title': 'Message', 'placeholder': "Content"},
  ]);
 dynamic response;
 
 @override
   Widget build(BuildContext context) {
     return Scaffold(
       appBar: AppBar(
         title: Text(widget.title),
       ),
       body: SingleChildScrollView(
         child: Container(
           child: Column(children: <Widget>[
             JsonWidget(
               form: form,
               onChanged: (dynamic response) {
                 this.response = response;
               },
             ),
             RaisedButton(
                 child: new Text('Send'),
                 onPressed: () {
                   print(this.response.toString());
                 })
           ]),
         ),
       ),
     );
   }
```

### When there is a change in the form, the (dynamic response;) is updated,
```
onChanged: (dynamic response) {
                 this.response = response;
               },
```

### when text is added to the TextField, add field called response
```
// initial form_send_email
[{"type":"Input","title":"Subject","placeholder":"Subject"},{"type":"TareaText","title":"Message","placeholder":"Content"}]

// add text (hi) in TextField Message, update dynamic response; and add field called response
[{type: Input, title: Subject, placeholder: Subject}, {type: TareaText, title: Message, placeholder: Content, response: hi }]
```


CHANGELOG.md
## [0.0.1] - 2022-01-18

* Project published

## [0.0.2] - 2022-01-18

* Documentation created

## [0.0.3] - 2022-01-18

* Documentation edit

## [0.0.4 & 0.0.4+1] - 2022-01-20

* Bugs resolved

lib/json_nav.dart
import 'dart:convert';

import 'package:flutter/material.dart';

class JsonNav extends StatefulWidget {
  const JsonNav({
    Key? key,
    required this.nav,
    this.padding,
    this.navMap,
    this.decorations = const {},
    this.buttonSave,
    this.actionSave,
  }) : super(key: key);

  final Map decorations;
  final String nav;
  final Map? navMap;
  final double? padding;
  final Widget? buttonSave;
  final Function? actionSave;

  @override
  _CoreNavState createState() => _CoreNavState(navMap ?? json.decode(nav));
}

class _CoreNavState extends State<JsonNav> {
  final dynamic navGeneral;

  // Return widgets

  List<Widget> jsonToBottomBar() {
    List<Widget> bottomWidgets = <Widget>[];

    for (int count = 0; count < navGeneral['fields'].length; count++) {
      Map item = navGeneral['fields'][count];

      if (item['type'] == 'text') {
        bottomWidgets.add(Text(item['bottomText']));
      }

      if(item.containsKey('iconName')){
        bottomWidgets.add(Text(item['iconName'],style: const TextStyle(fontFamily: 'MaterialIcons'),));
      }
    }

    return bottomWidgets;
  }

  List<Widget> jsonToNav() {
    List<Widget> navWidgets = <Widget>[];
    if (navGeneral['title'] != null) {
      navWidgets.add(Text(
        navGeneral['title'],
        style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
      ));
    }
    return navWidgets;
  }

  _CoreNavState(this.navGeneral);

  @override
  Widget build(BuildContext context) {
    return navGeneral['type'] == 'AppBar'
        ? AppBar(
      title: Text(navGeneral['title']),
      centerTitle: true,
    )
        : BottomAppBar(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: jsonToBottomBar(),
      ),
    );
  }
}


lib/json_form.dart
library json_to_widget;

import 'dart:convert';
import 'package:flutter/material.dart';

class JsonForm extends StatefulWidget {
  const JsonForm({Key? key,
    required this.form,
    required this.onChanged,
    this.padding,
    this.formMap,
    this.errorMessages = const {},
    this.validations = const {},
    this.decorations = const {},
    this.buttonSave,
    this.actionSave,
  }) : super(key: key);

  final Map errorMessages;
  final Map validations;
  final Map decorations;
  final String form;
  final Map? formMap;
  final double? padding;
  final Widget? buttonSave;
  final Function? actionSave;
  final ValueChanged<dynamic> onChanged;

  @override
  _CoreFormState createState() =>
      _CoreFormState(formMap ?? json.decode(form));
}

class _CoreFormState extends State<JsonForm> {
  final dynamic formGeneral;

  int? radioValue;

  // validators

  String isRequired(item, value) {
    if (value.isEmpty) {
      return widget.errorMessages[item['key']] ?? 'Please enter some text';
    }
    return value;
  }

  String validateEmail(item, String value) {
    String p = "[a-zA-Z0-9\+\.\_\%\-\+]{1,256}" +
        "\\@" +
        "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,64}" +
        "(" +
        "\\." +
        "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,25}" +
        ")+";
    RegExp regExp = RegExp(p);

    if (regExp.hasMatch(value)) {
      return value;
    }
    return 'Email is not valid';
  }

  bool labelHidden(item) {
    if (item.containsKey('hiddenLabel')) {
      if (item['hiddenLabel'] is bool) {
        return !item['hiddenLabel'];
      }
    } else {
      return true;
    }
    return false;
  }

  // Return widgets

  List<Widget> jsonToForm() {
    List<Widget> listWidget = <Widget>[];
    if (formGeneral['title'] != null) {
      listWidget.add(Text(
        formGeneral['title'],
        style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
      ));
    }
    if (formGeneral['description'] != null) {
      listWidget.add(Text(
        formGeneral['description'],
        style: const TextStyle(fontSize: 14.0,fontStyle: FontStyle.italic),
      ));
    }

    for (var count = 0; count < formGeneral['fields'].length; count++) {
      Map item = formGeneral['fields'][count];

      if (item['type'] == "Input" ||
          item['type'] == "Password" ||
          item['type'] == "Email" ||
          item['type'] == "TextArea" ||
          item['type'] == "TextInput") {
        Widget label = const SizedBox.shrink();
        if (labelHidden(item)) {
          label = Text(
            item['label'],
            style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0),
          );
        }

        listWidget.add(Container(
          margin: const EdgeInsets.only(top: 5.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              label,
              TextFormField(
                controller: null,
                initialValue:  formGeneral['fields'][count]['value'],
                decoration: item['decoration'] ??
                    widget.decorations[item['key']] ??
                    InputDecoration(
                      hintText: item['placeholder'] ?? "",
                      helperText: item['helpText'] ?? "",
                    ),
                maxLines: item['type'] == "TextArea" ? 10 : 1,
                onChanged: (String value) {
                  formGeneral['fields'][count]['value'] = value;
                  _handleChanged();
                },
                obscureText: item['type'] == "Password" ? true : false,
                validator: (value) {
                  if (widget.validations.containsKey(item['key'])) {
                    return widget.validations[item['key']](item, value);
                  }
                  if (item.containsKey('validator')) {
                    if (item['validator'] != null) {
                      if (item['validator'] is Function) {
                        return item['validator'](item, value);
                      }
                    }
                  }
                  if (item['type'] == "Email") {
                    return validateEmail(item, value!);
                  }

                  if (item.containsKey('required')) {
                    if (item['required'] == true ||
                        item['required'] == 'True' ||
                        item['required'] == 'true') {
                      return isRequired(item, value);
                    }
                  }

                  return null;
                },
              ),
            ],
          ),
        ));
      }

      if (item['type'] == "RadioButton") {
        List<Widget> radios = [];

        if (labelHidden(item)) {
          radios.add(Text(item['label'],
              style:
              const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0)));
        }
        radioValue = item['value'];
        for (var i = 0; i < item['items'].length; i++) {
          radios.add(
            Row(
              children: <Widget>[
                Expanded(
                    child: Text(
                        formGeneral['fields'][count]['items'][i]['label'])),
                Radio<int>(
                    value: formGeneral['fields'][count]['items'][i]['value'],
                    groupValue: radioValue,
                    onChanged: (int? value) {
                      setState(() {
                        radioValue = value!;
                        formGeneral['fields'][count]['value'] = value;
                        _handleChanged();
                      });
                    })
              ],
            ),
          );
        }

        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: radios,
            ),
          ),
        );
      }

      if (item['type'] == "Switch") {
        if (item['value'] == null) {
          formGeneral['fields'][count]['value'] = false;
        }
        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Row(children: <Widget>[
              Expanded(child: Text(item['label'])),
              Switch(
                value: item['value'] ?? false,
                onChanged: (bool value) {
                  setState(() {
                    formGeneral['fields'][count]['value'] = value;
                    _handleChanged();
                  });
                },
              ),
            ]),
          ),
        );
      }

      if (item['type'] == "Checkbox") {
        List<Widget> checkboxes = [];
        if (labelHidden(item)) {
          checkboxes.add(Text(item['label'],
              style:
              const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0)));
        }
        for (var i = 0; i < item['items'].length; i++) {
          checkboxes.add(
            Row(
              children: <Widget>[
                Expanded(
                    child: Text(
                        formGeneral['fields'][count]['items'][i]['label'])),
                Checkbox(
                  value: formGeneral['fields'][count]['items'][i]['value'],
                  onChanged: (bool? value) {
                    setState(
                          () {
                        formGeneral['fields'][count]['items'][i]['value'] =
                            value;
                        _handleChanged();
                      },
                    );
                  },
                ),
              ],
            ),
          );
        }

        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: checkboxes,
            ),
          ),
        );
      }

      if (item['type'] == "Select") {
        Widget label = const SizedBox.shrink();
        if (labelHidden(item)) {
          label = Text(item['label'],
              style:
              const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0));
        }

        listWidget.add(Container(
          margin: const EdgeInsets.only(top: 5.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              label,
              DropdownButton<String>(
                hint: const Text("Select a user"),
                value: formGeneral['fields'][count]['value'],
                onChanged: (String? newValue) {
                  setState(() {
                    formGeneral['fields'][count]['value'] = newValue;
                    _handleChanged();
                  });
                },
                items:
                item['items'].map<DropdownMenuItem<String>>((dynamic data) {
                  return DropdownMenuItem<String>(
                    value: data['value'],
                    child: Text(
                      data['label'],
                      style: const TextStyle(color: Colors.black),
                    ),
                  );
                }).toList(),
              ),
            ],
          ),
        ));
      }
    }

    if (widget.buttonSave != null) {
      listWidget.add(Container(
        margin: const EdgeInsets.only(top: 10.0),
        child: InkWell(
          onTap: () {
            if (_formKey.currentState!.validate()) {
              widget.actionSave!(formGeneral);
            }
          },
          child: widget.buttonSave,
        ),
      ));
    }
    return listWidget;
  }

  _CoreFormState(this.formGeneral);

  void _handleChanged() {
    widget.onChanged(formGeneral);
  }

  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Form(
      autovalidateMode: formGeneral['autoValidated'] ?? AutovalidateMode.onUserInteraction,
      key: _formKey,
      child: Container(
        padding: EdgeInsets.all(widget.padding ?? 8.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: jsonToForm(),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:json_to_widget/json_to_widget.dart';

void main() {
  runApp(HomePage());
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  String form = json.encode({
    'fields': [
      {
        'key': 'input1',
        'type': 'Input',
        'label': 'Username',
        'placeholder': "Enter Your Username",
        'required': true
      },
      {
        'key': 'password1',
        'type': 'Password',
        'label': 'Password',
        'required': true
      },
    ]
  });

  dynamic response;

  Map decorations = {
    'input1': InputDecoration(
        prefixIcon: Icon(Icons.account_box), border: OutlineInputBorder()),
    'password1': InputDecoration(
        prefixIcon: Icon(Icons.security), border: OutlineInputBorder())
  };

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Login"),
      ),
      body: Column(
        children: <Widget>[
          const Text(
            "Login Form",
            style: TextStyle(fontSize: 30.0, fontWeight: FontWeight.bold),
          ),
          JsonWidget(
            decorations: decorations,
            form: form,
            onChanged: (dynamic response) {
              this.response = response;
            },
            actionSave: (data) {
              print(data);
            },
            buttonSave: Container(
              height: 40.0,
              color: Colors.blueAccent,
              child: const Center(
                child: Text("Login",
                    style: TextStyle(
                        color: Colors.white, fontWeight: FontWeight.bold)),
              ),
            ),
          ),
        ],
      ),
    );
  }
}


lib/json_form.object.dart
library json_to_widget;

import 'dart:convert';
import 'package:flutter/material.dart';

import 'custom_decoration.dart';
import 'ron.dart';

class JsonFormObject extends StatefulWidget {
  const JsonFormObject({
    Key? key,
    required this.form,
    required this.onChanged,
    this.padding,
    this.formMap,
    this.errorMessages = const {},
    this.validations = const {},
    this.decorations = const {},
    this.buttonSave,
    this.actionSave,
  }) : super(key: key);

  final Map errorMessages;
  final Map validations;
  final Map decorations;
  final String form;
  final Map? formMap;
  final double? padding;
  final Widget? buttonSave;
  final Function? actionSave;
  final ValueChanged<dynamic> onChanged;

  @override
  _CoreWidgetState createState() =>
      _CoreWidgetState(formMap ?? json.decode(form));
}

class _CoreWidgetState extends State<JsonFormObject> {
  final dynamic widgetGeneral;

  int? radioValue;

  // validators

  String? isRequired(item, value) {
    if (value.isEmpty) {
      return widget.errorMessages[item['key']] ?? 'Lütfen zorunlu alanları doldurunuz';
    }
  }

  String? validateEmail(item, String value) {
    String p = "[a-zA-Z0-9\+\.\_\%\-\+]{1,256}" +
        "\\@" +
        "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,64}" +
        "(" +
        "\\." +
        "[a-zA-Z0-9][a-zA-Z0-9\\-]{0,25}" +
        ")+";
    RegExp regExp = RegExp(p);

    if (!regExp.hasMatch(value)) {
      return 'Email onaylı değil';
    }
  }

  bool labelHidden(item) {
    if (item.containsKey('hiddenLabel')) {
      if (item['hiddenLabel'] is bool) {
        return !item['hiddenLabel'];
      }
    } else {
      return true;
    }
    return false;
  }

  // Return widgets

  List<Widget> jsonToForm() {
    List<Widget> listWidget = <Widget>[];
    if (widgetGeneral['title'] != null) {
      listWidget.add(Text(
        widgetGeneral['title'],
        style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
      ));
    }
    if (widgetGeneral['description'] != null) {
      listWidget.add(Text(
        widgetGeneral['description'],
        style: const TextStyle(fontSize: 14.0, fontStyle: FontStyle.italic),
      ));
    }

    for (var count = 0; count < widgetGeneral['fields'].length; count++) {
      Map item = widgetGeneral['fields'][count];

      if (item.containsKey('command')) {
        if (item['command'].runtimeType == String) {
          Ron().call(item['command'], []);
        } else {
          Ron().call(item['command']['function'], {
            'args': item['command']['args'],
            'result': item['command']['result']
          });
        }
      }

      if (item['type'] == "container") {
        listWidget.add(Container(
            margin: EdgeInsets.only(
                top: item['top'] ?? 1,
                bottom: item['bottom'] ?? 1,
                left: item['left'] ?? 1,
                right: item['right'] ?? 1),
            width: item['width'],
            height: item['height'],
            color: Color.fromRGBO(
                item['R'], item['G'], item['B'], item['Opacity']),
            child: item['child'] == "text"
                ? Center(
              child: Text(
                item['text'],
              ),
            )
                : null));
      }

      if (item['type'] == "Input" ||
          item['type'] == "Password" ||
          item['type'] == "Email" ||
          item['type'] == "TextArea" ||
          item['type'] == "TextInput") {
        Widget label = const SizedBox.shrink();
        if (labelHidden(item)) {
          label = Text(
            item['label'],
            style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0),
          );
        }

        listWidget.add(Container(
          margin: const EdgeInsets.only(top: 5.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              label,
              TextFormField(
                controller: null,
                initialValue: widgetGeneral['fields'][count]['value'],
                decoration: item['decoration'] ??
                    widget.decorations[item['key']] ??
                    CustomDecoration().textFieldDecoration(item),
                maxLines: item['type'] == "TextArea" ? 10 : 1,
                onChanged: (String value) {
                  widgetGeneral['fields'][count]['value'] = value;
                  _handleChanged();
                },
                obscureText: item['type'] == "Password" ? true : false,
                validator: (value) {
                  if (widget.validations.containsKey(item['key'])) {
                    return widget.validations[item['key']](item, value);
                  }
                  if (item.containsKey('validator')) {
                    if (item['validator'] != null) {
                      if (item['validator'] is Function) {
                        return item['validator'](item, value);
                      }
                    }
                  }
                  if (item['type'] == "Email") {
                    return validateEmail(item, value!);
                  }

                  if (item.containsKey('required')) {
                    if (item['required'] == true ||
                        item['required'] == 'True' ||
                        item['required'] == 'true') {
                      return isRequired(item, value);
                    }
                  }

                  return null;
                },
              ),
            ],
          ),
        ));
      }

      if (item['type'] == "RadioButton") {
        List<Widget> radios = [];

        if (labelHidden(item)) {
          radios.add(Text(item['label'],
              style: const TextStyle(
                  fontWeight: FontWeight.bold, fontSize: 16.0)));
        }
        radioValue = item['value'];
        for (var i = 0; i < item['items'].length; i++) {
          radios.add(
            Row(
              children: <Widget>[
                Expanded(
                    child: Text(
                        widgetGeneral['fields'][count]['items'][i]['label'])),
                Radio<int>(
                    value: widgetGeneral['fields'][count]['items'][i]['value'],
                    groupValue: radioValue,
                    onChanged: (int? value) {
                      setState(() {
                        radioValue = value!;
                        widgetGeneral['fields'][count]['value'] = value;
                        _handleChanged();
                      });
                    })
              ],
            ),
          );
        }

        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: radios,
            ),
          ),
        );
      }

      if (item['type'] == "Switch") {
        if (item['value'] == null) {
          widgetGeneral['fields'][count]['value'] = false;
        }
        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Row(children: <Widget>[
              Expanded(child: Text(item['label'])),
              Switch(
                value: item['value'] ?? false,
                onChanged: (bool value) {
                  setState(() {
                    widgetGeneral['fields'][count]['value'] = value;
                    _handleChanged();
                  });
                },
              ),
            ]),
          ),
        );
      }

      if (item['type'] == "Checkbox") {
        List<Widget> checkboxes = [];
        if (labelHidden(item)) {
          checkboxes.add(Text(item['label'],
              style: const TextStyle(
                  fontWeight: FontWeight.bold, fontSize: 16.0)));
        }
        for (var i = 0; i < item['items'].length; i++) {
          checkboxes.add(
            Row(
              children: <Widget>[
                Expanded(
                    child: Text(
                        widgetGeneral['fields'][count]['items'][i]['label'])),
                Checkbox(
                  value: widgetGeneral['fields'][count]['items'][i]['value'],
                  onChanged: (bool? value) {
                    setState(
                          () {
                        widgetGeneral['fields'][count]['items'][i]['value'] =
                            value;
                        _handleChanged();
                      },
                    );
                  },
                ),
              ],
            ),
          );
        }

        listWidget.add(
          Container(
            margin: const EdgeInsets.only(top: 5.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: checkboxes,
            ),
          ),
        );
      }

      if (item['type'] == "Select") {
        Widget label = const SizedBox.shrink();
        if (labelHidden(item)) {
          label = Text(item['label'],
              style:
              const TextStyle(fontWeight: FontWeight.bold, fontSize: 16.0));
        }

        listWidget.add(Container(
          margin: const EdgeInsets.only(top: 5.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              label,
              DropdownButton<String>(
                hint: const Text("Select a user"),
                value: widgetGeneral['fields'][count]['value'],
                onChanged: (String? newValue) {
                  setState(() {
                    widgetGeneral['fields'][count]['value'] = newValue;
                    _handleChanged();
                  });
                },
                items:
                item['items'].map<DropdownMenuItem<String>>((dynamic data) {
                  return DropdownMenuItem<String>(
                    value: data['value'],
                    child: Text(
                      data['label'],
                      style: const TextStyle(color: Colors.black),
                    ),
                  );
                }).toList(),
              ),
            ],
          ),
        ));
      }
    }

    if (widget.buttonSave != null) {
      listWidget.add(Container(
        margin: const EdgeInsets.only(top: 10.0),
        child: InkWell(
          onTap: () {
            if (_formKey.currentState!.validate()) {
              widget.actionSave!(widgetGeneral);
            }
          },
          child: widget.buttonSave,
        ),
      ));
    }
    return listWidget;
  }

  _CoreWidgetState(this.widgetGeneral);

  void _handleChanged() {
    widget.onChanged(widgetGeneral);
  }

  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Form(
      autovalidateMode:
      widgetGeneral['autoValidated'] ?? AutovalidateMode.onUserInteraction,
      key: _formKey,
      child: Container(
        padding: EdgeInsets.all(widget.padding ?? 8.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: jsonToForm(),
        ),
      ),
    );
  }
}
